internal/remotestate/backend/s3/client_test.go

```diff
@@ -0,0 +1,218 @@
+//go:build aws
+
+package s3_test
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/gruntwork-io/terragrunt/internal/errors"
+	s3backend "github.com/gruntwork-io/terragrunt/internal/remotestate/backend/s3"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// defaultTestRegion is for simplicity, do all testing in the us-east-1 region
+const defaultTestRegion = "us-east-1"
+
+// CreateS3ClientForTest creates a DynamoDB client we can use at test time. If there are any errors creating the client, fail the test.
+func CreateS3ClientForTest(t *testing.T) *s3backend.Client {
+	t.Helper()
+
+	mockOptions, err := options.NewTerragruntOptionsForTest("aws_test")
+	require.NoError(t, err, "Error creating mockOptions")
+
+	extS3Cfg := &s3backend.ExtendedRemoteStateConfigS3{
+		RemoteStateConfigS3: s3backend.RemoteStateConfigS3{
+			Region: defaultTestRegion,
+		},
+	}
+
+	client, err := s3backend.NewClient(extS3Cfg, mockOptions)
+	require.NoError(t, err, "Error creating S3 client")
+
+	return client
+}
+
+func TestAwsCreateLockTableIfNecessaryTableDoesntAlreadyExist(t *testing.T) {
+	t.Parallel()
+
+	client := CreateS3ClientForTest(t)
+
+	WithLockTable(t, client, func(tableName string, client *s3backend.Client) {
+		AssertCanWriteToTable(t, tableName, client)
+	})
+}
+
+func TestAwsCreateLockTableConcurrency(t *testing.T) {
```

Comment:
Would a race test be more appropriate here?

Reply:
Not my code, let's do in next PR.